File: kotlin/src/main/kotlin/advertisements/advertisement/application/publishAdvertisement/PublishAdvertisementUseCase.kt
package advertisements.advertisement.application.publishAdvertisement

import advertisements.advertisement.domain.Advertisement
import advertisements.advertisement.domain.value_object.AdvertisementDate
import advertisements.advertisement.domain.value_object.AdvertisementId
import advertisements.advertisement.domain.value_object.Description
import advertisements.shared.value_object.CivicCenterId
import advertisements.shared.value_object.Email
import advertisements.shared.value_object.Password
import advertisements.user.domain.value_object.MemberNumber
import java.time.LocalDateTime

class PublishAdvertisementUseCase(private val advertisementRepository: advertisements.advertisement.domain.AdvertisementRepository) {
    fun execute(publishAdvertisementCommand: advertisements.advertisement.application.publishAdvertisement.PublishAdvertisementCommand) {
        val advertisementId = AdvertisementId(publishAdvertisementCommand.id)

        if (null !== advertisementRepository.findById(advertisementId)) {
            throw advertisements.advertisement.domain.exceptions.AdvertisementAlreadyExistsException.withId(advertisementId.value())
        }

        val advertisement = Advertisement(
            advertisementId,
            Description(publishAdvertisementCommand.description),
            Email(publishAdvertisementCommand.email),
            Password.fromPlainPassword(publishAdvertisementCommand.password),
            AdvertisementDate(LocalDateTime.now()),
            CivicCenterId.create(publishAdvertisementCommand.civicCenterId),
            MemberNumber(publishAdvertisementCommand.memberId)
        )

        advertisementRepository.save(advertisement)
    }
}

File: typescript/src/advertisements/advertisement/application/publish-advertisement/PublishAdvertisementUseCase.ts
import { AdvertisementRepository } from "../../domain/AdvertisementRepository";
import { Advertisement } from "../../domain/Advertisement";
import { PublishAdvertisementCommand } from "./PublishAdvertisementCommand";
import {Password} from "../../../shared/value-object/Password";
import {AdvertisementId} from "../../domain/value-object/AdvertisementId";
import {Description} from "../../domain/value-object/Description";
import {AdvertisementDate} from "../../domain/value-object/AdvertisementDate";
import {AdvertisementAlreadyExistsException} from "../../domain/exceptions/AdvertisementAlreadyExistsException";

export class PublishAdvertisementUseCase {

  constructor(
    private advertisementRepository: AdvertisementRepository
  ) {

  }

  async execute(command: PublishAdvertisementCommand): Promise<void> {
    const advertisementId = new AdvertisementId(command.id)

    if(await this.advertisementRepository.findById(advertisementId)) {
      throw AdvertisementAlreadyExistsException.withId(advertisementId.value())
    }

    const advertisement = new Advertisement(
      advertisementId,
      new Description(command.description),
      await Password.fromPlainPassword(command.password),
      new AdvertisementDate(new Date())
    )

    await this.advertisementRepository.save(advertisement)
  }
}
